### PR TITLE
chore: Validate before adding Xds resource

### PR DESCRIPTION
**Pull request by Ronnie-personal**
**What type of PR is this?**  Chore: Validate before adding Xds resource

**What this PR does / why we need it**: This PR calls Validate() before adding the resource to Xds resource table.

**Which issue(s) this PR fixes**: #1647 

Fixes #

### CODE REVIEW

**Comment by Ronnie-personal** [comment]
> thanks for raising this PR @Ronnie-personal . Imo attaching a `ValidateAll()` right before inserting the resource into `tCtx.AddXdsResource` is something the developer might forget to do, instead should we add `ValidateAll()` inside the `tCtx.AddXdsResource` and change the signature of it to return a error code instead ?

How about this code change?
func (t *ResourceVersionTable) AddXdsResource(rType resourcev3.Type, xdsResource types.Resource) error {
	// Perform type switch to handle different types of xdsResource
	switch r := xdsResource.(type) {
	case *listenerv3.Listener:
		// Handle Type specific operations
		if resourceOfType, ok := xdsResource.(*listenerv3.Listener); ok {
			if err := resourceOfType.ValidateAll(); err != nil {
				return fmt.Errorf("validation failed for xds resource %+v, err:%v", r, err)
			}
		} else {
			return fmt.Errorf("failed to cast xds resource %+v to Listener type", r)
		}
	case *routev3.RouteConfiguration:
		// Handle Type specific operations
		if resourceOfType, ok := xdsResource.(*routev3.RouteConfiguration); ok {
			if err := resourceOfType.ValidateAll(); err != nil {
				return fmt.Errorf("validation failed for xds resource %+v, err:%v", r, err)
			}
		} else {
			return errors.New("failed to cast xds resource to RouteConfiguration type", r)
		}

	case *Type3:
		// Handle Type3 specific operations

	// Add more cases for other types as needed
	default:
		// Handle the case when the type is not recognized or supported
		fmt.Errorf("Unsupported resource type: %v", rType)
	}

	if t.XdsResources == nil {
		t.XdsResources = make(XdsResources)
	}
	if t.XdsResources[rType] == nil {
		t.XdsResources[rType] = make([]types.Resource, 0, 1)
	}

	t.XdsResources[rType] = append(t.XdsResources[rType], xdsResource)
	return nil
}

**Comment by arkodg** [comment]
yeah @Ronnie-personal something similar to above logic is needed, you can probably use `rType resourcev3.Type` as in input to `switch` and then typecast from resource to the specific type before running `ValidateAll`

**Comment by Ronnie-personal** [comment]
@arkodg 
The e2e ratelimit test failed. Please find detailed error at https://github.com/envoyproxy/gateway/actions/runs/5633849037/job/15263275737?pr=1662#step:6:1913.

This PR does not change e2e test code. Do you think the error is caused by any code change in this pull request?

**Comment by Ronnie-personal** [comment]
@arkodg 

The existing tests failed.
I suspect the failure is caused by the ValidateAll() on the ClusterType.

Is it worth the effort to revert back the changes on ClusterType and test again?
Can you please advise on the next step?

File: internal/xds/types/resourceversiontable.go [hunk 1]
```diff
@@ -68,7 +74,60 @@ func (t *ResourceVersionTable) GetXdsResources() XdsResources {
...
+		if resourceOfType, ok := xdsResource.(*tlsv3.Secret); ok {
+			if err := resourceOfType.ValidateAll(); err != nil {
+				return fmt.Errorf("validation failed for xds resource %+v, err:%v", xdsResource, err)
+			}
+		} else {
+			return fmt.Errorf("failed to cast xds resource %+v to RouteConfiguration type", xdsResource)
+		}
+
+	case resourcev3.EndpointType:
+		// TBD - ValidateAll() breaks existing test internal/cmd/egctl/translate_test
```
Comment (arkodg) [inline internal/xds/types/resourceversiontable.go:110]: thanks for catching this, lets raise a issue to track this

File: internal/xds/types/resourceversiontable.go [hunk 2]
```diff
@@ -68,7 +74,60 @@ func (t *ResourceVersionTable) GetXdsResources() XdsResources {
...
+		if resourceOfType, ok := xdsResource.(*clusterv3.Cluster); ok {
+			if err := resourceOfType.ValidateAll(); err != nil {
+				return fmt.Errorf("validation failed for xds resource %+v, err:%v", xdsResource, err)
+			}
+		} else {
+			return fmt.Errorf("failed to cast xds resource %+v to RouteConfiguration type", xdsResource)
+		}
+	case resourcev3.RateLimitConfigType:
+		// Handle specific operations
+		// cfg resource from runner.go is the RateLimitConfig type from "github.com/envoyproxy/go-control-plane/ratelimit/config/ratelimit/v3", which does have validate function.
```
Comment (arkodg) [inline internal/xds/types/resourceversiontable.go:124]: this also raise a issue to track this

File: internal/xds/types/resourceversiontable_test.go [hunk 3]
```diff
@@ -233,7 +278,7 @@ func TestAddOrReplaceXdsResource(t *testing.T) {
 			name: "inject-new-listener",
 			tableIn: &ResourceVersionTable{
 				XdsResources: XdsResources{
```
Comment (arkodg) [inline internal/xds/types/resourceversiontable_test.go:280]: was this an existing bug ?

**Comment by arkodg** [comment]
LGTM @Ronnie-personal, this PR looks good, just added some minor comments, thanks again !

File: internal/xds/types/resourceversiontable_test.go (hunk 3 above)
Comment (Ronnie-personal) [reply internal/xds/types/resourceversiontable_test.go:280]: Thanks for the review.
I believe the code was carried over from the main branch when I created my branch.
I cannot recall making any changes to it.

But I agree that 'resourcev3.ListenerType: []types.Resource{},' makes more sense. We need to test adding a new listener to the table, so we wouldn't need to initialize 'tableIn' with the 'testListener'.

**Review by arkodg — APPROVED** [review]

File: internal/xds/types/resourceversiontable.go (hunk 2 above)
Comment (Ronnie-personal) [reply internal/xds/types/resourceversiontable.go:124]: Sure, I have opened https://github.com/envoyproxy/gateway/issues/1742. I will investigate further.

**Comment by Ronnie-personal** [comment]
@arkodg 
Sorry, I pushed a minor change to fix the typo.
Here are the changes https://github.com/Ronnie-personal/gateway/commit/d886fe3f053d70cd49d37e8ace62370affcf0a01

**Review by arkodg — APPROVED** [review]

**Review by zirain — APPROVED** [review]

File: internal/xds/types/resourceversiontable.go (hunk 1 above)
Comment (Ronnie-personal) [reply internal/xds/types/resourceversiontable.go:110]: Sure, I have opened https://github.com/envoyproxy/gateway/issues/1742. Here is the PR to fix the issue https://github.com/envoyproxy/gateway/pull/1758

**Review by Xunzhuo — APPROVED** [review]

**Pull request merged**